python/cuml/decomposition/pca_mg.pyx

```diff
@@ -0,0 +1,605 @@
+        arr_interfaces_trans = []
+                    
+        if self.dtype == np.float32:
+            data = self._build_dataFloat(arr_interfaces)
+            arr_interfaces_trans = self._build_transData(partsToRanks, rnk, self.n_components, np.float32)
+            trans_data = self._build_dataFloat(arr_interfaces_trans)
+
+            fit_transform(handle_[0],
+                <RankSizePair**>rankSizePair,
+                <size_t> n_total_parts,
+                <floatData_t**> data,
+                <floatData_t**> trans_data,
+                <float*> comp_ptr,
+                <float*> explained_var_ptr,
+                <float*> explained_var_ratio_ptr,
+                <float*> singular_vals_ptr,
+                <float*> mean_ptr,
+                <float*> noise_vars_ptr,
+                params,
+                True)
+        else:
+            data = self._build_dataDouble(arr_interfaces)
+            arr_interfaces_trans = self._build_transData(partsToRanks, rnk, self.n_components, np.float64)
+            trans_data = self._build_dataDouble(arr_interfaces_trans)
+            
+            fit_transform(handle_[0],
+                <RankSizePair**>rankSizePair,
+                <size_t> n_total_parts,
+                <doubleData_t**> data,
+                <doubleData_t**> trans_data,
+                <double*> comp_ptr,
+                <double*> explained_var_ptr,
+                <double*> explained_var_ratio_ptr,
+                <double*> singular_vals_ptr,
+                <double*> mean_ptr,
+                <double*> noise_vars_ptr,
+                params,
+                True)
+              
+        self.handle.sync()
+                     
+            # make sure the previously scheduled gpu tasks are complete before the
+        # following transfers start
+        
+        for idx in range(n_total_parts):
+            free(<RankSizePair*>rankSizePair[idx])
+        free(<RankSizePair**>rankSizePair)
+
+        # Keeping the additional dataframe components during cuml 0.8.
+        # See github issue #749
+        self.components_ = cudf.DataFrame()
+        for i in range(0, params.n_cols):
+            n_c = params.n_components
+            self.components_[str(i)] = self.components_ary[i*n_c:(i+1)*n_c]
+
+        if (isinstance(X, cudf.DataFrame)):
+            del(X_m)
+        
+        trans_cudf = []         
+ 
+        if _transform:
+            for x_i in arr_interfaces_trans:
+                trans_cudf.append(cudf.DataFrame.from_gpu_matrix(x_i["obj"]))
+
+            if self.dtype == np.float32:
+                self._freeFloatD(trans_data, arr_interfaces_trans)    
+                self._freeFloatD(data, arr_interfaces)        
+            else:
+                self._freeDoubleD(trans_data, arr_interfaces_trans)
+                self._freeDoubleD(data, arr_interfaces)
+
+            return trans_cudf
+    
+
+    def transform(self, X, M, N, partsToRanks, rnk):
```

Comment:
Has something changed w/ the implementation of transform and inverse transform? I thought this was embarrassingly parallel. Once the parameters like the singular values are computed, why aren't we able to use the single-GPU version for this?